doc/tutorials/computing-product-analytics.md

```diff
@@ -0,0 +1,258 @@
+
++++
+
+### Dataset Description
+
+For simplicity, we just use some small-scale randomly generated data. The dataset **user_activity** is consists of three columns: 
+- User_id : the unique identifier of the user
+- Month: the month where 
+- Activity_count: the number of activity conduct by the user in that month. If the user never used this app before this month, this is considered as their sign-up month. 
+
+
+| user_id | date       | activity_count |
+|--------|------------|----------------|
+| 1      | 2021-01-01 | 5              |
+| 1      | 2021-02-01 | 3              |
+| 1      | 2021-03-01 | 2              |
+| 2      | 2021-01-01 | 10             |
+| 3      | 2021-02-01 | 1              |
+| 3      | 2021-03-01 | 0              |
+| 4      | 2021-02-01 | 6              |
+| 5      | 2021-01-01 | 4              |
+| 5      | 2021-02-01 | 5              |
+| 5      | 2021-03-01 | 6              |
+| 6      | 2021-03-01 | 7              |
+| 7      | 2021-03-01 | 10             |
+
+```{code-cell} ipython3
+:tags: [hide-output]
+
+%pip install jupysql duckdb-engine --quiet
+```
+
+```{code-cell} ipython3
+%load_ext sql
+%sql duckdb://
+```
+
+Let's create the table by ourself first using SQL.
+
+```{code-cell} ipython3
+:tags: [hide-output]
+
+%%sql
+DROP TABLE IF EXISTS user_activity;
+CREATE TABLE user_activity (
+  user_id INT NOT NULL,
+ date DATE NOT NULL,
+  activity_count INT NOT NULL,
+  PRIMARY KEY (user_id, date)
+);
+INSERT INTO user_activity (user_id, date, activity_count)
+VALUES
+  (1, '2021-01-01', 5),
+  (1, '2021-02-01', 3),
+  (1, '2021-03-01', 2),
+  (2, '2021-01-01', 10),
+  (3, '2021-02-01', 1),
+  (3, '2021-03-01', 0),
+  (4, '2021-02-01', 6),
+  (5, '2021-01-01', 4),
+  (5, '2021-02-01', 5),
+  (5, '2021-03-01', 6),
+  (6, '2021-03-01', 7),
+  (7, '2021-03-01', 10);
+```
+
+Let's check our table using select.
+
+```{code-cell} ipython3
+:tags: [hide-output]
+
+%%sql
+SELECT * FROM user_activity
+```
+
+### Growth 
+
+We defined growth rate to be the percentage increase of total number of users between each month. 
+
+We first retrive the total number of users in each month
+
+```{code-cell} ipython3
+:tags: [hide-output]
+
+%%sql
+Select MONTH(date) as month, COUNT(DISTINCT user_id) AS total_users
+FROM user_activity
+GROUP BY MONTH(date)
+```
+
+Here, we group the dataset by the month of the date, then count the number of distinct users as our total number of users. This command will return the following output: 
+
+| month | total_users       |
+|--------|------------|
+| 1      | 3|
+| 2      | 4 |
+| 3      | 5 |
+
++++
+
+We save it using a temporary table called CTE. Then calculate using self join. 
+
+```{tip}
+'/' in sql between two integers conduct integer division only. For example, 10/3 would be 3 instead of 3.33333. As a result, we multiply the formula by 1.0 to convert it to float division. 
+```
+
+```{code-cell} ipython3
+:tags: [hide-output]
+
+%%sql
+WITH CTE AS(
```

Comment:
we have a CTE feature that can help you make this more readable: https://jupysql.ploomber.io/en/latest/compose.html